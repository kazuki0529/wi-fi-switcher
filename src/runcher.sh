#!/bin/bash

set -eux

###########################################################
# 引数で渡された評価対象の日付が期間内か判定する
#
# @param target 評価対象の日付(yyyy-MM-dd hh:mm:ss)
# @param period_start 開始日(yyyy-MM-dd hh:mm:ss)
# @param period_end 終了日(yyyy-MM-dd hh:mm:ss)
#
# @return 期間内の場合は0を返し、期間外は1を返す
###########################################################
is_within_period() {
  declare -r target=$1
  [[ "${2}" = "" ]] \
    && declare -r period_start='1990-01-01 00:00:00' \
    || declare -r period_start=$2
  [[ "${3}" = "" ]] \
    && declare -r period_end='1990-01-01 00:00:00' \
    || declare -r period_end=$3

  # 評価対象が期間内か判定する
  declare -r start_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_start" +%s))
  declare -r end_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_end" +%s))
  if [ $start_difference -ge 0 ] && [ $end_difference -le 0 ]; then
    return 0;
  else
    return 1;
  fi
}

declare -r now=$1
declare -r schedule=$2

cd $(dirname $0)

while IFS=, read start end description; do
  # ヘッダ行判定
  [[ "${start}" == "start" ]] && continue

  is_within_period "${now}" "${start}" "${end}" \
    && (yarn cypress run -s ./cypress/integration/PR-500KI-wifi-on.ts || true) \
    &&  exit 0
done < "${schedule}"

yarn cypress run -s ./cypress/integration/PR-500KI-wifi-off.ts
