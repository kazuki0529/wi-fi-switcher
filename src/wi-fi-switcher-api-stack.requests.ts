import { APIGatewayEvent, APIGatewayProxyResult } from 'aws-lambda';
import moment from 'moment';
import RequestUseCase from './usecase/RequestUseCase';

export async function handler(
  event: APIGatewayEvent,
): Promise<APIGatewayProxyResult> {
  const usecase = RequestUseCase();

  const pathPart = event.resource.substring('/v1'.length);

  switch (pathPart) {
    case '/requests': {
      if (event.httpMethod === 'POST') {
        const body = event.body ? JSON.parse(event.body) : undefined;
        if (!body) throw new Error('The Body section is empty.');

        const req = {
          ...body,
          start: moment(body.start),
          end: moment(body.end),
        };

        const data = await usecase.create(req);
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data),
        };
      }
      if (event.httpMethod === 'GET') {
        const data = await usecase.getAll();
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data.map(e => ({
            ...e,
            start: e.start.toISOString(),
            end: e.end.toISOString(),
            createdAt: e.createdAt?.toISOString(),
            updatedAt: e.updatedAt?.toISOString(),
          }))),
        };
      }

      throw new Error('An undefined method has been called.');
    };
    case '/requests/{requestId}': {
      if (event.httpMethod !== 'PUT') throw new Error('An undefined method has been called.');

      const id = event.pathParameters ? event.pathParameters.requestId : undefined;
      if (!id) throw new Error('The id has not been specified.');

      const body = event.body ? JSON.parse(event.body) : undefined;
      if (!body) throw new Error('The Body section is empty.');

      const req = {
        ...body,
        start: body.start ? moment(body.start) : undefined,
        end: body.end ? moment(body.end) : undefined,
      };

      const data = await usecase.update(id, req);
      return {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json; charset=utf-8',
        },
        body: JSON.stringify(data),
      };
    };
    default: {
      throw new Error('An undefined path has been called.');
    };
  }
}
