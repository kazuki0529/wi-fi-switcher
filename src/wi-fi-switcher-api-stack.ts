import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamo from '@aws-cdk/aws-dynamodb';
import * as iam from '@aws-cdk/aws-iam';;
import * as lambda from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';

import { Construct, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';

export type StackStage = 'staging' | 'prod';

export class WiFiSwitcherApiStack extends Stack {
  public readonly table: dynamo.Table;
  public readonly api: apigateway.RestApi;

  constructor(
    scope: Construct, id: string,
    props?: StackProps,
  ) {
    super(scope, id, props);

    // データ格納領域作成
    this.table =new dynamo.Table(this, 'data', {
      partitionKey: {
        name: '_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_s_key',
        type: dynamo.AttributeType.STRING,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: dynamo.BillingMode.PAY_PER_REQUEST,
    });
    this.table.addGlobalSecondaryIndex({
      indexName: 'GSI-1',
      partitionKey: {
        name: '_gsi1_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_gsi1_s_key',
        type: dynamo.AttributeType.STRING,
      },
      projectionType: dynamo.ProjectionType.ALL,
    });

    /**
     * API作成
     */
    const requests = new NodejsFunction(this, 'requests', {
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_14_X,
      timeout: cdk.Duration.seconds(60),
      memorySize: 2048,
      environment: {
        REGION: this.region,
        TABLE_NAME: this.table.tableName,
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
      role: new iam.Role(this, 'generateBatchLambdaRole', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AWSLambdaBasicExecutionRole',
          ),
        ],
      }),
    });
    this.table.grantReadWriteData(requests);

    this.api = new apigateway.RestApi(this, 'game-play', {
      policy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            principals: [new iam.AnyPrincipal()],
            effect: iam.Effect.ALLOW,
            actions: ['execute-api:Invoke'],
            resources: ['execute-api:/prod/*'],
          }),
        ],
      }),
      deployOptions: {
        tracingEnabled: true,
        stageName: 'prod',
      },
    });
    const baseApi = this.api.root.addResource('v1');
    const requestsIntegration = new apigateway.LambdaIntegration(requests);
    const requestsApi = baseApi.addResource('requests');
    requestsApi.addMethod('POST', requestsIntegration);
    requestsApi.addMethod('GET', requestsIntegration);

    requestsApi.addResource('{requestId}').addMethod('PUT', requestsIntegration);

    new cdk.CfnOutput(this, 'ApiUrl', { value: this.api.url });
  }
}