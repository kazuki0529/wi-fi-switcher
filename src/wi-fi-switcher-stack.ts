import * as apigateway from '@aws-cdk/aws-apigatewayv2';
import * as certificate from '@aws-cdk/aws-certificatemanager';
import * as cloudfront from '@aws-cdk/aws-cloudfront';
import * as iam from '@aws-cdk/aws-iam';
import { ARecord, PublicHostedZone, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { Construct, Stack, StackProps } from '@aws-cdk/core';


export type StackStage = 'staging' | 'prod';

interface WiFiSwitcherStackProps extends StackProps {
  readonly api: apigateway.HttpApi;
  readonly apiStage: apigateway.HttpStage;
  readonly stage: StackStage;
}

export class WiFiSwitcherStack extends Stack {
  public readonly webBucket: s3.Bucket;
  public readonly distribution: cloudfront.CloudFrontWebDistribution;

  constructor(
    scope: Construct, id: string,
    props: WiFiSwitcherStackProps,
  ) {
    super(scope, id, props);

    /**
     * Webコンテンツのリリース先バケット
    **/
    this.webBucket = new s3.Bucket(this, 'web-bucket', {
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      encryption: s3.BucketEncryption.S3_MANAGED,
    });
    /**
     * CloudFrontを経由しないアクセスは拒否りたいので、OAIを作成する
    **/
    const oai = new cloudfront.OriginAccessIdentity(this, 'web-oai', {
      comment: 's3 access.',
    });
    /**
     * 作成したOAIからのアクセスを許可するポリシー
    **/
    const policy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['s3:GetObject'],
      principals: [new iam.CanonicalUserPrincipal(oai.cloudFrontOriginAccessIdentityS3CanonicalUserId)],
      resources: [
        this.webBucket.bucketArn + '/*',
      ],
    });
    this.webBucket.addToResourcePolicy(policy);


    /**
     * CloudFront
    **/
    const hostedZoneId = this.node.tryGetContext('ZONE_ID');
    const zoneName = this.node.tryGetContext('ZONE_NAME');
    const domain = `${(props.stage === 'staging') ? 'dev.': ''}${zoneName}`;
    const fqdn = `game-play.home.${domain}`;
    const certArn = this.node.tryGetContext('CERT_ARN');
    const cert = certArn
      ? certificate.Certificate.fromCertificateArn(this, 'CertificateForTheDomain', certArn)
      : undefined;
    this.distribution = new cloudfront.CloudFrontWebDistribution(this, 'cfn-distribution', {
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.webBucket,
            originAccessIdentity: oai,
          },
          behaviors: [{ isDefaultBehavior: true }],
        },
        {
          customOriginSource: {
            domainName: `${props.api.apiId}.execute-api.${this.region}.${this.urlSuffix}`,
          },
          behaviors: [
            {
              pathPattern: `${props.apiStage.stageName}/*`,
              allowedMethods: cloudfront.CloudFrontAllowedMethods.ALL,
              forwardedValues: {
                headers: ['Content-Type', 'Accept', 'Accept-Encoding', 'Accept-Language', 'Authorization', 'Origin'],
                queryString: true,
              },
              defaultTtl: cdk.Duration.seconds(0),
              maxTtl: cdk.Duration.seconds(0),
              minTtl: cdk.Duration.seconds(0),
            },
          ],
        },
      ],
      priceClass: cloudfront.PriceClass.PRICE_CLASS_200,
      viewerCertificate: cert
        ? cloudfront.ViewerCertificate.fromAcmCertificate(cert, {
          aliases: [fqdn],
        })
        : {
          aliases: [],
          props: {
            cloudFrontDefaultCertificate: true,
          },
        },
    });

    /**
     * Route53へAレコードへ登録
     */
    if (hostedZoneId && zoneName && fqdn) {
      const zone = PublicHostedZone.fromHostedZoneAttributes(
        this,
        'register-public-host-zone',
        {
          hostedZoneId,
          zoneName,
        },
      );
      new ARecord(this, 'register-a-record', {
        zone,
        recordName: fqdn,
        target: RecordTarget.fromAlias(
          new CloudFrontTarget(this.distribution),
        ),
      });
    }

    // CloudFrontのアクセスURLを出力
    new cdk.CfnOutput(this, 'UiCloudFrontUrl', {
      value: `https://${this.distribution.distributionDomainName}/`,
    });
    new cdk.CfnOutput(this, 'UiAccessUrl', {
      value: `https://${fqdn}/`,
    });
  }
}