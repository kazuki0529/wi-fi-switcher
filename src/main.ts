import { App } from '@aws-cdk/core';
import { WiFiSwitcherStack } from './wi-fi-switcher-stack';


// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new WiFiSwitcherStack(app, 'wi-fi-switcher-stack', { env: devEnv });

app.synth();
