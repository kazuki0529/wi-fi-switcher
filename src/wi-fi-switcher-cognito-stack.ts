import * as cognito from '@aws-cdk/aws-cognito';
import * as cdk from '@aws-cdk/core';

import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { StackStage } from './wi-fi-switcher-stack';

interface WiFiSwitcherCognitoStackProps extends StackProps {
  readonly stage: StackStage;
}

export class WiFiSwitcherCognitoStack extends Stack {
  readonly userPool: cognito.UserPool;
  readonly userPoolClient: cognito.UserPoolClient;

  constructor(
    scope: Construct, id: string,
    props: WiFiSwitcherCognitoStackProps,
  ) {
    super(scope, id, props);

    const zoneName = this.node.tryGetContext('ZONE_NAME');
    const domain = `${(props.stage === 'staging') ? 'dev.': ''}${zoneName}`;
    const fqdn = `game-play.home.${domain}`;

    this.userPool = new cognito.UserPool(this, 'userPool', {
      selfSignUpEnabled: false,
      standardAttributes: {
        email: { required: true, mutable: true },
        phoneNumber: { required: false },
      },
      signInCaseSensitive: false,
      autoVerify: { email: true },
      signInAliases: { username: true },
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    this.userPool.addDomain('domain', {
      cognitoDomain: { domainPrefix: fqdn.toLowerCase().split('.').join('-') },
    });

    this.userPoolClient = this.userPool.addClient('client', {
      oAuth: {
        scopes: [
          cognito.OAuthScope.EMAIL,
          cognito.OAuthScope.OPENID,
          cognito.OAuthScope.PROFILE,
        ],
        callbackUrls: [`https://${fqdn}/`],
        logoutUrls: [`https://${fqdn}/`],
        flows: { authorizationCodeGrant: true },
      },
    });
  }
}