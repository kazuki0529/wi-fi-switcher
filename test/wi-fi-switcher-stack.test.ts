import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { WiFiSwitcherApiStack } from '../src/wi-fi-switcher-api-stack';
import { WiFiSwitcherStack } from '../src/wi-fi-switcher-stack';

test('Snapshot', () => {
  const app = new App();
  const api = new WiFiSwitcherApiStack(app, 'api');
  const stack = new WiFiSwitcherStack(app, 'test', { api: api.api });

  expect(stack).toHaveResource('AWS::CloudFront::Distribution');
  expect(stack).toHaveResource('AWS::CloudFront::CloudFrontOriginAccessIdentity');
  expect(stack).toHaveResource('AWS::S3::Bucket');

  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});
