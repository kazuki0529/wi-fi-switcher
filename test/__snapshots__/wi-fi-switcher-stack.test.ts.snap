// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "UiAccessUrl": Object {
      "Value": "https://undefined/",
    },
    "UiCloudFrontUrl": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Fn::GetAtt": Array [
                "cfndistributionCFDistributionBEDBC4FD",
                "DomainName",
              ],
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601ArtifactHashBD3FCB86": Object {
      "Description": "Artifact hash for asset \\"abb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601\\"",
      "Type": "String",
    },
    "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3Bucket284F96CE": Object {
      "Description": "S3 bucket for asset \\"abb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601\\"",
      "Type": "String",
    },
    "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3VersionKey18236449": Object {
      "Description": "S3 key for asset version \\"abb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601\\"",
      "Type": "String",
    },
    "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0ArtifactHash9D8F179A": Object {
      "Description": "Artifact hash for asset \\"e3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0\\"",
      "Type": "String",
    },
    "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0S3BucketB3DDCC13": Object {
      "Description": "S3 bucket for asset \\"e3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0\\"",
      "Type": "String",
    },
    "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0S3VersionKey3418DF70": Object {
      "Description": "S3 key for asset version \\"e3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68ArtifactHashD9A515C3": Object {
      "Description": "Artifact hash for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7": Object {
      "Description": "S3 bucket for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F": Object {
      "Description": "S3 key for asset version \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0S3BucketB3DDCC13",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0S3VersionKey3418DF70",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse3d9996b6fafcc7da88312672e15e3cc925b02cffc6f01a615d81f22303e3ae0S3VersionKey3418DF70",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "s3deployAwsCliLayer46C4636C",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.7",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3Bucket284F96CE",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3Bucket284F96CE",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "webbucket77AC4F5E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "webbucket77AC4F5E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "cfndistributionCFDistributionBEDBC4FD": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [],
          "CacheBehaviors": Array [
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "DefaultTTL": 0,
              "ForwardedValues": Object {
                "Headers": Array [
                  "Content-Type",
                  "Accept",
                  "Accept-Encoding",
                  "Accept-Language",
                  "Authorization",
                  "Origin",
                ],
                "QueryString": true,
              },
              "MaxTTL": 0,
              "MinTTL": 0,
              "PathPattern": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::ImportValue": "api:ExportsOutputRefgameplayDeploymentStageapiE6CF920D69456EB3",
                    },
                    "/*",
                  ],
                ],
              },
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "webbucket77AC4F5E",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "weboai463C5C5E",
                      },
                    ],
                  ],
                },
              },
            },
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "https-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::ImportValue": "api:ExportsOutputRefgameplayCFF713D3EF165255",
                    },
                    ".execute-api.",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ".",
                    Object {
                      "Ref": "AWS::URLSuffix",
                    },
                  ],
                ],
              },
              "Id": "origin2",
            },
          ],
          "PriceClass": "PriceClass_200",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "s3deployAwsCliLayer46C4636C": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "s3deployCustomResource1CD3FAE7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "webbucket77AC4F5E",
        },
        "DistributionId": Object {
          "Ref": "cfndistributionCFDistributionBEDBC4FD",
        },
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3Bucket284F96CE",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3VersionKey18236449",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersabb3bd7dff29c41d61752db4f963df1709f8b0fab699aa779ac57f0352404601S3VersionKey18236449",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "webbucket77AC4F5E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "aws-cdk:cr-owned:c946f39f",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "webbucketPolicy99027222": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "webbucket77AC4F5E",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "weboai463C5C5E",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "webbucket77AC4F5E",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "weboai463C5C5E": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "s3 access.",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;
